tsdb/engine/tsm1/cache.go

```diff
@@ -0,0 +1,320 @@
+package tsm1
+
+import (
+	"container/list"
+	"fmt"
+	"sort"
+	"sync"
+
+	"github.com/influxdb/influxdb/tsdb"
+)
+
+var ErrCacheMemoryExceeded = fmt.Errorf("cache maximum memory size exceeded")
+var ErrCacheInvalidCheckpoint = fmt.Errorf("invalid checkpoint")
+
+// lru orders string keys from least-recently used to most-recently used. It is not
+// goroutine safe.
+type lru struct {
+	list     *list.List
+	elements map[string]*list.Element
+}
+
+// newLRU returns an initialized LRU.
+func newLRU() *lru {
+	return &lru{
+		list:     list.New(),
+		elements: make(map[string]*list.Element),
+	}
+}
+
+// MoveToFront marks key as the most recently used key.
+func (l *lru) MoveToFront(key string) {
+	e, ok := l.elements[key]
+	if !ok {
+		l.elements[key] = l.list.PushFront(key)
+		return
+	}
+	l.list.MoveToFront(e)
+}
+
+// Remove removes key from the LRU. If the key does not exist nothing happens.
+func (l *lru) Remove(key string) {
+	if _, ok := l.elements[key]; ok {
+		l.list.Remove(l.elements[key])
+		delete(l.elements, key)
+	}
+}
+
+// Front returns the most-recently used key. If there is no such key, then "" is returned.
+func (l *lru) Front() string {
+	e := l.list.Front()
+	if e == nil {
+		return ""
+	}
+	return e.Value.(string)
+}
+
+// Back returns the least-recently used key. If there is no such key, then "" is returned.
+func (l *lru) Back() string {
+	e := l.list.Back()
+	if e == nil {
+		return ""
+	}
+	return e.Value.(string)
+}
+
+// DoFromLeast iterates through the LRU, from least-recently used to most-recently used,
+// calling the given function with each key.
+func (l *lru) DoFromLeast(f func(key string)) {
+	for e := l.list.Back(); e != nil; e = e.Prev() {
+		f(e.Value.(string))
+	}
+}
+
+// entry is the set of all values received for a given key.
+type entry struct {
+	values   Values // All stored values.
```

Comment:
This struct also looks like it need a mutex since you could have potentially have to `add` calls concurrently on the same key/checkpoint.